src/trace.rs

```diff
@@ -0,0 +1,309 @@
+const ENV_LOG: &str = "LINKERD2_PROXY_LOG";
+
+use linkerd2_error::Error;
+use std::{env, fmt, str, time::Instant};
+use tokio_timer::clock;
+use tracing::{info, Dispatch, Event, Level};
+use tracing_subscriber::{
+    filter,
+    fmt::{format, Builder, Context, Formatter},
+};
+use tracing_subscriber::{reload, EnvFilter, FmtSubscriber};
+
+type SubscriberBuilder = Builder<format::NewRecorder, Format, filter::LevelFilter>;
+type Subscriber = Formatter<format::NewRecorder, Format>;
+
+#[derive(Clone)]
+pub struct LevelHandle {
+    inner: reload::Handle<EnvFilter, Subscriber>,
+}
+
+/// Initialize tracing and logging with the value of the `ENV_LOG`
+/// environment variable as the verbosity-level filter.
+pub fn init() -> Result<LevelHandle, Error> {
+    let env = env::var(ENV_LOG).unwrap_or_default();
+    let (dispatch, handle) = with_filter(env);
+
+    // Set up log compatibility.
+    init_log_compat()?;
+    // Set the default subscriber.
+    tracing::dispatcher::set_global_default(dispatch)?;
+    Ok(handle)
+}
+
+pub fn init_log_compat() -> Result<(), Error> {
+    tracing_log::LogTracer::init().map_err(Error::from)
+}
+
+pub fn with_filter(filter: impl AsRef<str>) -> (Dispatch, LevelHandle) {
+    let filter = filter.as_ref();
+
+    // Set up the subscriber
+    let builder = subscriber_builder()
+        .with_env_filter(filter)
+        .with_filter_reloading();
+    let handle = LevelHandle {
+        inner: builder.reload_handle(),
+    };
+    let dispatch = Dispatch::new(builder.finish());
+
+    (dispatch, handle)
+}
+
+/// Returns a builder that constructs a `FmtSubscriber` that logs trace events.
+fn subscriber_builder() -> SubscriberBuilder {
+    let start_time = clock::now();
+    FmtSubscriber::builder().on_event(Format { start_time })
+}
+
+struct Format {
+    start_time: Instant,
+}
+
+impl<N> tracing_subscriber::fmt::FormatEvent<N> for Format
```

Comment:
This custom implementation was really only necessary in order to format the custom log context along with the `tracing` span. Since we're no longer doing that, this can be replaced with the default implementation (with ANSI formatting disabled). If we want to keep the current formatting for timestamps, we'd need to provide a custom [timestamp formatter](https://docs.rs/tracing-subscriber/0.1.5/tracing_subscriber/fmt/time/trait.FormatTime.html) (although there is an [`Uptime` formatter](https://docs.rs/tracing-subscriber/0.1.5/tracing_subscriber/fmt/time/struct.Uptime.html), it formats the timestamps differently from ours).

Reply:
:+1: let's address in a followup